dep0/lib/05_llvmgen/src/gen_func.cpp

```diff
@@ -0,0 +1,128 @@
+#include "private/gen_func.hpp"
+
+#include "private/gen_alloca.hpp"
+#include "private/gen_attrs.hpp"
+#include "private/gen_body.hpp"
+#include "private/gen_type.hpp"
+
+#include <llvm/IR/Attributes.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/Function.h>
+
+#include <string>
+
+namespace dep0::llvmgen {
+
+static void gen_func_args(local_context_t&, llvm_func_proto_t const&, llvm::Function*);
+static void gen_func_attributes(local_context_t const&, llvm_func_proto_t const&, llvm::Function*);
+static void gen_func_body(
+    global_context_t&,
+    local_context_t const&,
+    llvm_func_proto_t const&,
+    typecheck::body_t const&,
+    llvm::Function*);
+
+void gen_func_args(local_context_t& local, llvm_func_proto_t const& proto, llvm::Function* const llvm_f)
+{
+    auto llvm_arg_it = llvm_f->arg_begin();
+    if (is_alloca_needed(proto.ret_type()))
+    {
+        assert(llvm_f->arg_size() == proto.args().size() + 1ul and "function with sret must have 1 more argument");
+        llvm_arg_it->addAttr(llvm::Attribute::StructRet);
+        llvm_arg_it->addAttr(llvm::Attribute::NonNull);
+        ++llvm_arg_it;
+    }
+    else
+        assert(llvm_f->arg_size() == proto.args().size() and "function and prototype must have same arguments");
+    for (auto const& arg: proto.args())
+    {
+        auto const& llvm_arg = llvm_arg_it++;
```

Comment:
no dereference here? maybe they are stored in a vector? doesn't matter should still deref and take address where necessary